File: 프로그래머스/카카오2022인턴십/3 - 코딩 테스트 공부.cpp
#include <bits/stdc++.h>
using namespace std;
int d[155][155];

int solution(int alp, int cop, vector<vector<int>> problems) {
    //왜 목표 초기를 0으로하면 틀리고 alp, cop로 하면 맞을까?
    int alp_target = alp, cop_target = cop;
    for(auto p : problems){
        alp_target = max(alp_target,p[0]);
        cop_target = max(cop_target,p[1]);
    }
    cout << alp_target << " " << cop_target;
    for(int i=0;i<155;i++) fill(d[i],d[i]+155,0x7f7f7f7f);
    problems.push_back({0,0,1,0,1});
    problems.push_back({0,0,0,1,1});
    d[alp][cop] = 0;
    for(int a = alp;a<=alp_target;a++){
        for(int c = cop;c<=cop_target;c++){
            for(auto problem : problems){
                int alp_req = problem[0];
                int cop_rep = problem[1];
                int alp_rwd = problem[2];
                int cop_rwd = problem[3];
                int cost = problem[4];
                if(a < alp_req || c < cop_rep) continue;
                int nxa = alp_rwd + a;
                int nxc= cop_rwd + c;
                if(nxa > alp_target) nxa = alp_target;
                if(nxc > cop_target) nxc = cop_target;
                d[nxa][nxc] = min(d[nxa][nxc], d[a][c] + cost);
            }
        }
    }
    return d[alp_target][cop_target];
}

File: 프로그래머스/카카오2022인턴십/3 - 코딩 테스트 공부.java
import java.util.*;
class Solution {
    int[][] dp;
    public int solution(int alp, int cop, int[][] problems) {
        int targetA = alp, targetC = cop;
        for(int[] problem : problems){
            targetA = Math.max(targetA, problem[0]);
            targetC = Math.max(targetC, problem[1]);
        }
        if(alp == targetA && cop == targetC) return 0;
        dp = new int[targetA * 2 + 10][targetC * 2 + 10];
        for(int i=0;i<=targetA * 2; i++) Arrays.fill(dp[i],0x7f7f7f7f);
        for(int i=0;i<=alp;i++){
            for(int j=0;j<=cop;j++){
                dp[i][j] = 0;
            }
        }
        for(int a = alp; a < targetA * 2 - 1; a++){
            for(int c = cop; c < targetC * 2 - 1; c++){
                dp[a + 1][c] = Math.min(dp[a][c] + 1, dp[a+1][c]);
                dp[a][c+1] = Math.min(dp[a][c] + 1, dp[a][c+1]);
                for(int[] problem : problems){
                    int alp_req = problem[0];
                    int cop_req = problem[1];
                    int alp_rwd = problem[2];
                    int cop_rwd = problem[3];
                    int cost = problem[4];
                    if(a < alp_req || c < cop_req) continue;
                    int nextA = a + alp_rwd;
                    int nextC = c + cop_rwd;
                    if(nextA > targetA) nextA = targetA;
                    if(nextC > targetC) nextC = targetC;
                    dp[nextA][nextC] = Math.min(dp[a][c] + cost, dp[nextA][nextC]);
                }
            }
        }
    
        return dp[targetA][targetC];
    }
}
